### DISCUSSION TITLE
iOS App + Custom CA = SSL Error

### DISCUSSION THREAD

**Discussion by DaCHack** [dani-garcia/vaultwarden#1446] [Q&A]
### Subject of the issue
I did a lot of research and found a lot of hints but anyways did not get my iOS app to work.
My iOS 14.4 on iPad Pro does not accept the server's certificate even though the root CA was added as a .p12-profile.
On Windows 10 / Google Chrome everything works like a charm.
Any chance that you can help me fix this? Or am I fighting a bug here?

### Your environment
* Bitwarden_rs version: 2.18.1 installed as docker image on Raspberry OS
* Clients used: iOS Bitwarden client, Safari
* Reverse proxy and version: none

### Steps to reproduce
- Clean install of bitwarden via docker image on host "server" with IP 192.168.1.4 on port 8000
- Root CA created via XCA
   (SHA-512,
   X509v3 Key Usage critical: Digital Signature, Key Encipherment, Data Encipherment, Key Agreement, Certificate Sign, CRL Sign
   X509v3 Extended Key Usage critical: TLS Web Server Authentication)
- Create server certificate via XCA
  (X509v3 Extended Key Usage: TLS Web Server Authentication
   X509v3 Subject Alternative Name: IP Address:192.168.1.1, IP Address:192.168.1.2, IP Address:192.168.1.3, IP Address:192.168.1.4, DNS:*.fritz.box, DNS:server.fritz.box
- Both certificates with limited validity to only 2 years as the Apple requirements state
- Try to connect to https://192.168.1.4:8000 or https://server.fritz.box:8000 via Safari, Bitwarden iOS app or Windows/Chrome

### Expected behaviour
- Webinterface shows with valid certificate
- iOS connects to the vault without issues
Some websites indicate that one should see the imported root CA in iOS under "General" -> "Info" ->"Certificate Trust..." but for me it is not visible there. Other websites indicate that this is not needed anymore as long as the CA is showing as "validated" under "General" --> "Profiles" (which is true for me)

### Actual behaviour
- Webinterface shows invalid certificate
- iOS shows SSL error when trying to login

**Comment by BlackDex** [comment]
Well we have a custom cert wiki which works for at least browsers, never tried an ios device. Though we do not support self-signed certificates and also not recommend using this. Neither is this an issue with bitwarden it self.

I suggest reading https://github.com/dani-garcia/bitwarden_rs/wiki/Running-a-private-bitwarden_rs-instance-with-Let%27s-Encrypt-certs
And
https://github.com/dani-garcia/bitwarden_rs/wiki/Private-CA-and-self-signed-certs-that-work-with-Chrome

Also, it could very well be that these kind of self-signed certs just don't work in these devices. Also, some applications and devices do not allow self-signed certificates which have a validity of longer the n days for example or some other reasons.

**Reply by DaCHack** [reply]
Many thanks for the quick response, especially given the fact that the guidance says that this approach is unsupported.
Highly appreciated!

Still, I struggle to understand the value for most end users if I need to expose my Bitwarden server directly or indirectly via a proxy to the internet. Wouldnt it be much more secure if I could use encrypted HTTPS communication with my local server but also only provide access to this server from within the local network?
Do you really go the public route to access the bitwarden vault? Or is there any other approach that I just do not see at the moment?

Many thanks!

Edit: The validity aspect I already considered and changed the initial post accordingly.
Other option: Do the hash algorithms of the CA and the server need to match? I have sha-512 for the CA and just tried to adapt my server cert to the discription in your guidance (sha-256) but did not want to replace the whole CA just for testing...

**Comment by BlackDex** [comment]
The biggest reason we suggest to use a proxy has two reasons.
1. SSL support. The library we use for exposing the endpoints to be available via http(s) is relatively young, and has some issues with some clients which are better handled by other software more focused on this.
2. That same http(s) library doesn't support websockets (yet). For this we need to use an other library and tcp port to have that supported. But we can't merge those two on the server side. That is where the proxy takes it over and handles these requests.

You don't need to expose the server to the internet for a valid acme cert. If i understand caddy correctly it can communicate with some dns providers to authenticate the cert requests without having your server exposed.

**Reply by DaCHack** [reply]
OK, got that. Many thanks.

I thought I could just use lego from your guidance above and retrieve the Let's encrypt certificates without using caddy inbetween.

I tried to copy the cert and key files to the input folder for the bitwarden docker container.
But when I start bitwarden with
`sudo docker run -d --name bitwarden -e ROCKET_TLS='{certs="/ssl/certs.pem",key="/ssl/key.pem"}' -v /etc/lighttpd/certs/:/ssl/ -v /home/dockerd/:/data/ -p 8000:80 --restart always bitwardenrs/server:latest`

The docker logs gives me:
`[2021-02-27 15:17:44.008][rocket::config::error][ERROR] environment variable ROCKET_TLS={certs="/ssl/certs.pem",key="/ssl/key.pem"} could not be parsed`

Is it impossible to let Rocket-TLS use these certificates?

**Reply by jjlin** [reply]
You need to request RSA certs instead. I updated https://github.com/dani-garcia/bitwarden_rs/wiki/Running-a-private-bitwarden_rs-instance-with-Let%27s-Encrypt-certs#getting-certs-using-the-lego-cli with the details.

**Reply by DaCHack** [reply]
I solved the issue.
It seems rocket cannot handle ec256 encryption.
Adding the parameter "-k rsa4096" solves the issue and the container runs without errors.

Thanks for pushing me in the right direction. Sometimes you just need to explain the problem to someone to uncover the solution :)

**Reply by DaCHack** [reply]
> You need to request RSA certs instead. I updated https://github.com/dani-garcia/bitwarden_rs/wiki/Running-a-private-bitwarden_rs-instance-with-Let%27s-Encrypt-certs#getting-certs-using-the-lego-cli with the details.

Just 30sek late :) But many thanks, mate!

**Reply by haydonryan** [reply]
> The biggest reason we suggest to use a proxy has two reasons.
> 
>     1. SSL support. The library we use for exposing the endpoints to be available via http(s) is relatively young, and has some issues with some clients which are better handled by other software more focused on this.
> 
>     2. That same http(s) library doesn't support websockets (yet). For this we need to use an other library and tcp port to have that supported. But we can't merge those two on the server side. That is where the proxy takes it over and handles these requests.

2 years on - I'm wondering if the two reasons are still valid? Does rocket_ws cover the websockets use case?

**Reply by BlackDex** [reply]
It does.

**Comment by mbrksntrk** [comment]
I resolved the problem by using server certificate for not more than 398 days. (CA is doesn't matter, can be 10 years)

Signature Digest: RSA-SHA256
SAN: DNS: <should be same with vaultwarden/bitwarden service eg:vaultwarden.home>
KU: Digital Signature, Key Encipherment
EKU: TLS Web Server Authentication, TLS Web Client Authentication, IP Security IKE Intermediate
Key Type: RSA
Key Size: 2048
Total Lifetime: 398 days
Lifetime Remaining: 397 days until expiration
